File: multiboot.py
import os
import TileWorldClient
import time
import math

#https://github.com/maciel310/gba-mmo-proxy/blob/main/multiboot.c


def multiboot(epIn, epOut, path):
    content = 0
    content = bytearray(open(path, 'rb').read())
    fsize = os.path.getsize(path)
    if fsize > 0x40000:
        print("File size error, max 256KB")
        exit()

    recv = 0
    while True:
        TileWorldClient.send(0x6202, epOut)
        recv = TileWorldClient.readall(epIn)
        if (recv >> 16) == 0x7202:
            break
    print("Lets do this thing!")
    TileWorldClient.send(0x6102, epOut)

    for i in range(96):
        out = (int(content[(i*2)])) + (int(content[(i*2)+1]) << 8)
        TileWorldClient.send(out, epOut)

    TileWorldClient.send(0x6200, epOut)
    TileWorldClient.send(0x6200, epOut)
    TileWorldClient.send(0x63D1, epOut)
    #Clear buffer
    TileWorldClient.readall(epIn)

    TileWorldClient.send(0x63D1, epOut)
    token = TileWorldClient.readall(epIn)
    if (token >> 24) != 0x73:
        print("Failed handshake!")
        exit()
    else:
        print("Handshake successful!")

    crcA = (token >> 16) & 0xFF
    seed = 0xFFFF00D1 | (crcA << 8)
    crcA = (crcA + 0xF) & 0xFF

    TileWorldClient.send((0x6400 | crcA), epOut)
    TileWorldClient.readall(epIn)

    fsize += 0xF
    fsize &= ~0xF

    TileWorldClient.send((fsize - 0x190) // 4, epOut)
    token = TileWorldClient.readall(epIn)
    crcB = (token >> 16) & 0xFF
    crcC = 0xC387
    print(fsize)
    print("Sending data!")
    for i in range(0xC0, fsize, 4):
        dat = int(content[i])
        dat += int(content[i + 1]) << 8
        dat += int(content[i + 2]) << 16
        dat += int(content[i + 3]) << 24

        tmp = dat

        for b in range(32):
            bit = (crcC ^ tmp) & 1
            if bit == 0:
                crcC = (crcC >> 1) ^ 0
            else:
                crcC = (crcC >> 1) ^ 0xc37b
            tmp >>= 1

        seed = seed * 0x6F646573 + 1
        dat = seed ^ dat ^ (0xFE000000 - i) ^ 0x43202F2F

        TileWorldClient.send(dat & 0xFFFFFFFF, epOut, True)

    print("Data sent")
    TileWorldClient.readall(epIn)

    tmp = 0xFFFF0000 | (crcB << 8) | crcA

    for b in range(32):
        bit = (crcC ^ tmp) & 1
        if bit == 0:
            crcC = (crcC >> 1) ^ 0
        else:
            crcC = (crcC >> 1) ^ 0xc37b
        tmp >>= 1

    TileWorldClient.send(0x0065, epOut)
    while True:
        TileWorldClient.send(0x0065, epOut)
        recv = TileWorldClient.readall(epIn)
        if (recv >> 16) == 0x0075:
            break

    TileWorldClient.send(0x0066, epOut)
    TileWorldClient.send(crcC & 0xFFFF, epOut)
    print("DONE!")

File: Pico Firmware/multiboot.c
#include "multiboot.h"
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include "pico/stdlib.h"

//Read 4 bytes and write four bytes
int rw4(uint32_t data)
{
    //TODO: This
    return 0;
}
//REFERENCE: http://problemkaputt.de/gbatek-bios-multi-boot-single-game-pak.htm
int multiboot(uint8_t* rom, uint64_t len)
{
    if(len > 0x40000)
    {
        printf("*** MULTIBOOT FILESIZE ERROR ***\n");
        return 0; //Failed
    }
    uint32_t temp;

    printf("Turn on the GBA\n");
    while((temp>>16) != 0x7202)
    {
        //Repeat 15 times, if failed: delay 1/16s and restart
        for(uint32_t i = 0; i < 15; i++)
        {
            temp = rw4(0x6200);
            if((temp>>16) == 0x7202)
                break;
            if(i == 14) sleep_ms(62.5);
        }
    }
    printf("Let's do this thing!\n");
    
    rw4(0x6102);

    //Send Nintendo logo (totally legal, I have the copyright ah yes)
    for(int i = 0; i < 96; i++)
        rw4((uint32_t)(rom[i*2]) + ((uint32_t)(rom[(i*2)+1]) << 8));
    
    rw4(0x6200);
    rw4(0x6202);
    rw4(0x63D1);

    temp = rw4(0x63D1);
    if((temp >> 24) != 0x73)
    {
        printf("Failed handshake! How sad\n");
        return 0;
    }
    printf("Handshake successful!\n");

    sleep_ms(62.5);

    uint32_t crcA = (temp >> 16) & 0xFF;
    uint32_t seed = 0xFFFF00D1 | (crcA << 8);
    crcA = (crcA + 0xF) & 0xFF;

    rw4(0x6400 | crcA);

    len += 0xF;
    len &= ~0xF;

    temp = rw4((len - 0x190) / 4);
    uint32_t crcB = (temp >> 16) % 0xFF;
    uint32_t crcC = 0xC387;

    uint32_t* rom4 = (uint32_t*)rom;
    uint32_t dat;
    uint32_t bit;
    for(int i = 0xC0; i < len; i += 4)
    {
        dat = rom4[i/4];
        temp = dat;

        for(int j = 0; j<32; j++)
        {
            bit = (crcC ^ temp) & 1;
            if(bit == 0)
                crcC = (crcC >> 1) ^ 0;
            else
                crcC = (crcC >> 1) ^ 0xc37b;
            temp >>= 1;
        }
        seed = seed * 0x6F646573 + 1;
        dat = seed ^ dat ^ (0xFE000000 - i) ^ 0x43202F2F;

        rw4(dat & 0xFFFFFFFF);
    }

    temp = 0xFFFF0000 | (crcB << 8) | crcA;

    for(int j = 0; j<32; j++)
    {
        bit = (crcC ^ temp) & 1;
        if(bit == 0)
            crcC = (crcC >> 1) ^ 0;
        else
            crcC = (crcC >> 1) ^ 0xc37b;
        temp >>= 1;
    }

    rw4(0x0065);

    while((rw4(0x0065) >> 16) != 0x0075);

    rw4(0x0066);
    rw4(crcC & 0xFFFF);

    printf("~~~~~~~~~~~~~~~~~~~~~~~~~~DONE!~~~~~~~~~~~~~~~~~~~~~~~~~~\n");
    return 1;
}